src/django_otp/models.py

```diff
@@ -226,10 +225,28 @@ class SideChannelDevice(Device):
         default=timezone.now,
         help_text="The timestamp of the moment of expiry of the saved token.",
     )
+    last_generated = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text="The last time a token was generated for this device.",
+    )
 
     class Meta:
         abstract = True
 
+    def generate_is_allowed(self):
```

Comment:
I think this stub could live in the root Device model, alongside verify_is_allowed. This would also simplify OTPAuthenticationFormMixin.

Reply:
It would be more tidy to have it in the same place as `verify_is_allowed`. However,  I thought to place it in `SideChannelDevice` because `generation_is_allowed` only applies only to the devices that generate token.

Reply:
I am not expert on the subject, I don't know what other channels are usually used to deliver a token, apart from email or SMS. But from my point of view, both email and sms should be throttled. But you are the expert on the subject so let me know if this is not the case and how to move this forward.

Reply:
> I don't know what other channels are usually used to deliver a token, apart from email or SMS.

This is a good logical fallacy to become familiar with: it's sometimes called proof by lack of imagination.

This is a framework. There are an infinite number of devices that might be written and the only rule is that they must inherit from `Device`. If you're thinking just in terms of the specific devices you know about or imagine, then you're thinking too small. The question is what's the right thing to do for every device that has ever or will ever be written.

Mixins and things like `SideChannelDevice` are convenience classes that help people write standard categories of devices in standard ways. But we can never imagine all possibilities, so any device authors are free to ignore the helpers and start from scratch. That's why `generate_token` and `verify_is_allowed` are on the root `Device` class: they don't apply to *every* device, but they could apply to *any* device.

Reply:
I considered `SideChannelDevice` as the class to subclass in case someone needed a device that uses token generation, verification and expiration.



> Mixins and things like `SideChannelDevice` are convenience classes that help people write standard categories of devices in standard ways. But we can never imagine all possibilities, so any device authors are free to ignore the helpers and start from scratch. That's why `generate_token` and `verify_is_allowed` are on the root `Device` class: they don't apply to _every_ device, but they could apply to _any_ device.

I see your point and agree. As you suggested, the `generate_is_allowed` should be in the `Device`.